firmware/port.c

```diff
@@ -248,6 +253,72 @@ void uart_send_data(PortData *p){
     }
 }
 
+uint16_t analog_read(Pin p) {
+    // switch pin mux to analog in
+    pin_adc(p);
+
+    // disable adc
+    ADC->CTRLA.reg &= ~ADC_CTRLA_ENABLE;
+
+    // set up clock
+    PM->APBCMASK.reg |= PM_APBCMASK_ADC;
+
+    // enable clock adc channel
+    gclk_enable(GCLK_ADC, GCLK_SOURCE_DFLL48M, 1);
+    GCLK->CLKCTRL.reg = GCLK_CLKCTRL_CLKEN |
+        GCLK_CLKCTRL_GEN(GCLK_ADC) |
+        GCLK_CLKCTRL_ID(GCLK_ADC_ID);
+
+    ADC->INPUTCTRL.reg = (ADC_INPUTCTRL_MUXPOS(p.adc) // select from proper pin
+        | ADC_INPUTCTRL_MUXNEG_GND // 0 = gnd
+        | ADC_INPUTCTRL_GAIN_DIV2); // gain of 1/2
+
+    // divide prescaler by 512 (93.75KHz), max adc freq is 2.1MHz
+    ADC->CTRLB.reg = ADC_CTRLB_PRESCALER_DIV512;
+    ADC->REFCTRL.reg = ADC_REFCTRL_REFSEL_INTVCC1; // reference voltage is 1/2 VDDANA
+    ADC->SWTRIG.reg = ADC_SWTRIG_START; // start adc
+    // wait until synced
+    while(ADC->STATUS.reg & ADC_STATUS_SYNCBUSY);
+
+    // enable
+    ADC->CTRLA.reg = ADC_CTRLA_ENABLE;
```

Comment:
Split out the one-time initialization into a function `adc_enable` in `common/` and call it from `main`